### ISSUE TITLE
Enum map does not play nice with ValueType

### ISSUE THREAD

**Issue by lptr** [HaxeFoundation/haxe#2479]
This code works as expected:

``` haxe
import Type;

class Test2 {
    static function main(){
        var map:Map<ValueType,String> = new Map();
        map.set(TInt, "TInt");
        map.set(TClass(String), "TClass(String)");
        map.set(TClass(Color), "TClass(Color)");
        map.set(TClass(Length), "TClass(Length)");
        map.set(TFloat, "TFloat");

        trace("TInt: " + map.get(TInt));
        trace("TFloat: " + map.get(TFloat));
        trace("TClass(String): " + map.get(TClass(String)));
        trace("TClass(Color): " + map.get(TClass(Color)));
        trace("TClass(Length): " + map.get(TClass(Length)));
    }
}

class Color {
    public function clone() {
        return this;
    }
}

class Length {
    public function clone() {
        return this;
    }
}
```

However, if I move `map.set(TClass(Color), "TClass(Color)");` one line up, the code does not find `TClass(Color)` anymore:

``` haxe
import Type;

class Test2 {
    static function main(){
        var map:Map<ValueType,String> = new Map();
        map.set(TInt, "TInt");
        map.set(TClass(Color), "TClass(Color)");  // <-- moved one line higher
        map.set(TClass(String), "TClass(String)");
        map.set(TClass(Length), "TClass(Length)");
        map.set(TFloat, "TFloat");

        trace("TInt: " + map.get(TInt));
        trace("TFloat: " + map.get(TFloat));
        trace("TClass(String): " + map.get(TClass(String)));
        trace("TClass(Color): " + map.get(TClass(Color)));
        trace("TClass(Length): " + map.get(TClass(Length)));
    }
}

class Color {
    public function clone() {
        return this;
    }
}

class Length {
    public function clone() {
        return this;
    }
}
```

This prints this on the console:

``` text
[Log] TInt: TInt (run, line 37)
[Log] TFloat: TFloat (run, line 38)
[Log] TClass(String): TClass(String) (run, line 39)
[Log] TClass(Color): null (run, line 45)
[Log] TClass(Length): TClass(Length) (run, line 51)
```

Essentially, not finding `TClass(Color)` anymore.

**Comment by romamik** [comment]
> Actually there's a real problem with objects and comparisons on js:
> 
> ```
> var x = {};
> var y = {};
> x == y; // false
> x < y; //false
> x > y; //false
> x >= y; //true!?
> x <= y; //true!?
> ```
> 
> So I guess this is something we can/should deal with when comparing Dynamics

Yes, EnumValueMap does not work correctly if key has object parameter. This test fails on js, but runs fine on cpp. I was going to report this as new issue but found this issue. I think it should be reopened.
```haxe
class FooParam {
	public function new() {
	}
}

enum EFoo {
	Foo(a: FooParam);
}

class Main {

	static function assert(v: Bool, m: String) {
		if (!v) {
			throw m;
		}
	}

	static function main() {

		var map: Map<EFoo, Int> = [];
		var keys: Array<EFoo> = [for (i in 0...100) EFoo.Foo(new FooParam())];
		for (i in 0...keys.length) {
			map[keys[i]] = i;
		}
		for (i in 0...keys.length) {
			assert(map[keys[i]] == i, 'map[keys[i]] == i, i=$i');
		}
	}
}
```

**Comment by RealyUniqueName** [comment]
Looking at `EnumValueMap` I don't understand why does it extend `BalancedTree` and use `Reflect.compare`. 
We have `Type.enumEq`, which works perfectly fine for any enum values. 
If anything `EnumValueMap` should have target-specific implementations like `ObjectMap` instead of a generic `BalancedTree`.

**Comment by back2dos** [comment]
We want `EnumValueMap` to satisfy `map[Some(42)] == map[Some(42)]`, which is why `ObjectMap` won't do the trick. And so we're left with using a tree, to not have everything run at linear cost and more.

**Comment by RealyUniqueName** [comment]
`Type.enumEq(Some(42), Some(42))` emits `true`.
And I didn't mean to use the same implementations `ObjectMap` use. I meant implementations should be platform specific.
Yes, probably at some map size it would be slower. But it would still be correct and it would work for any `EnumValue`.

**Comment by zommerfelds** [comment]
I'm not sure if this issue is unique to JS. I was trying to build a `Map<Class<T>, Foo>` and tried to use `ValueType` as an alternative to `Class<T>`. It could be that two different `ValueType.TClass(type)` instances were not being compared to equal and I inserted two elements with the same type.

This was on HashLink. If you are interested I could investigate more.

**Comment by zommerfelds** [comment]
Sorry, it turns out that the issue is I was using Anonymous Structures as the Map key (as a pair class), but this doesn't work as two equal copies of a structure don't compare to be the same key in ObjectMap. Ignore my comment above.

**Comment by RealyUniqueName** [comment]
I'd like to resolve this issue by merging #9670 
Any objections?

**Comment by yz-xlame** [comment]
Have we made any progress on this matter

**Comment by RblSb** [comment]
> Seems like something is wrong with JS way of comparing Class (which are created with functions), which makes it unsuitable for such sorting. You should consider using the class name instead.
> 
> ```
> Color = function() {}
> function () {}
> Length = function() {}
> function () {}
> Color > Length;
> false
> Color < Length;
> false
> Color >= Length;
> true
> Color <= Length;
> true
> Color == Length;
> false
> ```

If problem is still with this, we can generate classes and methods as `Color = function Color() {}` for correct comparison.
Maybe this can also improve debugging in some cases, see https://github.com/HaxeFoundation/haxe/issues/6085

**Comment by Simn** [comment]
Closed by #12139.

Yes I know this took over 11 years, but bugfixes are like wine or something.
